src/Symfony/Component/HttpFoundation/JsonResponse.php

```diff
@@ -171,6 +171,11 @@ public function setEncodingOptions($encodingOptions)
     {
         $this->encodingOptions = (int) $encodingOptions;
 
+        // If callback is not used, try to update $this->data with decoded content
+        if ($this->callback === null) {
+            return $this->setData(json_decode($this->content));
```

Comment:
The magic and mixed responsibilities were already there.
I think we need proper solution for this, I can have a go at it.